mono/metadata/object.c

```diff
@@ -5062,15 +5071,29 @@ mono_string_new_utf32 (MonoDomain *domain, const mono_unichar4 *text, gint32 len
 MonoString *
 mono_string_new_size (MonoDomain *domain, gint32 len)
 {
+	MonoError error;
+	MonoString *str = mono_string_new_size_checked (domain, len, &error);
+	mono_error_raise_exception (&error);
+
+	return str;
+}
+
+MonoString *
+mono_string_new_size_checked (MonoDomain *domain, gint32 len, MonoError *error)
+{
 	MONO_REQ_GC_UNSAFE_MODE;
 
 	MonoString *s;
 	MonoVTable *vtable;
 	size_t size;
 
+	mono_error_init (error);
+
 	/* check for overflow */
-	if (len < 0 || len > ((SIZE_MAX - G_STRUCT_OFFSET (MonoString, chars) - 8) / 2))
-		mono_gc_out_of_memory (-1);
+	if (len < 0 || len > ((SIZE_MAX - G_STRUCT_OFFSET (MonoString, chars) - 8) / 2)) {
+		mono_error_set_out_of_memory (error, "Could not allocate %i bytes", -1);
+		return NULL;
+	}
 
 	size = (G_STRUCT_OFFSET (MonoString, chars) + (((size_t)len + 1) * 2));
 	g_assert (size > 0);
```

Comment:
We already check that size would not overflow. So if it actually does, we should assert.